packages/testing/src/execution_testing/fixtures/pre_alloc_groups.py

```diff
@@ -86,15 +108,177 @@ def to_file(self, file: Path) -> None:
+    group when it is accessed.
+
+    Also holds a cached model dump of the pre-allocation group, either in
+    string or JSON format depending on the last request.
+    """
+
+    _pre_alloc_group: "PreAllocGroup" = PrivateAttr(init=False)
+    _model_dump_cache: ModelDumpCache | None = PrivateAttr(None)
+    _cache_miss_count: int = PrivateAttr(0)
+
+    def state_root(self) -> Hash:
+        """On pre-alloc groups, which are normally very big, we always cache."""
+        return self._pre_alloc_group.genesis.state_root
+
+    def model_dump(  # type: ignore[override]
+        self, mode: Literal["json", "python"], **kwargs: Any
+    ) -> Any:
+        """
+        Model dump the pre-allocation group, with caching.
+
+        Note: 'mode' here follows Pydantic's semantics:
+        - 'python' -> standard model_dump
+        - 'json'   -> JSON-compatible python data
+        """
+        if (
+            self._model_dump_cache is not None
+            and self._model_dump_cache.model_dump_mode == mode
+            and self._model_dump_cache.model_dump_type == "dict"
+            and self._model_dump_cache.model_dump_config == kwargs
+        ):
+            return self._model_dump_cache.data
+
+        self._cache_miss_count += 1
+        data = super().model_dump(mode=mode, **kwargs)
+        self._model_dump_cache = ModelDumpCache(
+            model_dump_mode=mode,
+            model_dump_config=kwargs,
+            model_dump_type="dict",
+            data=data,
+        )
+        return data
+
+    def model_dump_json(self, **kwargs: Any) -> str:
+        """Model dump the pre-allocation group in JSON string format, with caching."""
+        if (
+            self._model_dump_cache is not None
+            and self._model_dump_cache.model_dump_mode == "json"
+            and self._model_dump_cache.model_dump_type == "string"
+            and self._model_dump_cache.model_dump_config == kwargs
+        ):
+            return self._model_dump_cache.data
+
+        self._cache_miss_count += 1
+        data = super().model_dump_json(**kwargs)
+        self._model_dump_cache = ModelDumpCache(
+            model_dump_mode="json",
+            model_dump_config=kwargs,
+            model_dump_type="string",
+            data=data,
+        )
+        return data
+
+
+class PreAllocGroup(PreAllocGroupBuilder):
+    """
+    Pre-allocation group for tests with identical Environment and fork values.
+
+    Groups tests by a hash of their fixture Environment and fork to enable
+    pre-allocation group optimization.
+    """
+
+    # Allow both field names and aliases
+    model_config = {"populate_by_name": True}
+    pre: GroupPreAlloc
+    genesis: FixtureHeader
+    pre_account_count: int
+    test_count: int
+
+    def model_post_init(self, __context: Any) -> None:
+        """
+        Model post init method to set GroupPreAlloc reference.
+        """
+        super().model_post_init(__context)
+        self.pre._pre_alloc_group = self
```

Comment:
I got it after a mo, but it's kinda obfuscated. Suggestion here https://github.com/marioevz/execution-specs/pull/2